### PR TITLE
Export topologies #20435

### PR DESCRIPTION
**Pull request by demonodojo**

### CODE REVIEW

File: sinfonier-web/app/models/module/ModuleVersion.java [hunk 1]
```diff
@@ -261,44 +261,49 @@ public void recheck() throws SinfonierException {
...
+			dir.mkdirs();
+		}
+
+		String fileName = module.getName() + ".json";
+		File file = FileUtils.getFile(dir, fileName);
+
+		try {
+			FileUtils.writeByteArrayToFile(file, res.toString().getBytes("UTF-8"));
+		} catch (IOException e) {
+			Logger.error(e.getMessage());
```
Comment (ivan11paths) [inline sinfonier-web/app/models/module/ModuleVersion.java:278]: If an error happens this code only print a log. You should return an error.

File: sinfonier-web/app/models/responses/Codes.java [hunk 2]
```diff
@@ -5,6 +5,7 @@
 public enum Codes {
   CODE_200("OK", 200),
   CODE_400("NOOK", 400),
```
Comment (ivan11paths) [inline sinfonier-web/app/models/responses/Codes.java]: You can add correct text for all codes.

**Comment by amatosc** [comment]
Do **not** merge to Master!!!

**Review by ivan11paths — CHANGES_REQUESTED** [review]

**Review by fandrod — APPROVED** [review]

File: sinfonier-web/app/models/topology/Wire.java [hunk 3]
```diff
@@ -8,7 +8,7 @@
 import com.google.gson.annotations.SerializedName;
 
 public class Wire {
-  private String xType;
+  private String xtype;
```
Comment (amatosc) [inline sinfonier-web/app/models/topology/Wire.java:11]: Why do you need rename it?

File: sinfonier-web/app/models/topology/json/serializers/TopologySerializer.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    {
+    	JsonObject jsonModule = new JsonObject();
+    	jsonModule.addProperty(FIELD_NAME, topologyModule.getName());
+    	jsonModule.addProperty(models.SinfonierConstants.TopologyModule.FIELD_VERSION_CODE, topologyModule.getVersionCode());
+    	jsonModule.addProperty(models.SinfonierConstants.TopologyModule.FIELD_TYPE, topologyModule.getVersionCode());
+    	jsonModule.addProperty(models.SinfonierConstants.TopologyModule.FIELD_LANGUAGE, topologyModule.getVersionCode());
+    	jsonModule.addProperty(models.SinfonierConstants.TopologyModule.FIELD_PARALLELISMS, topologyModule.getVersionCode());
+    	jsonModule.add(models.SinfonierConstants.TopologyModule.FIELD_VALUES, parser.parse(gson.toJson(topologyModule.getValues())).getAsJsonObject());
+    	jsonModule.add(models.SinfonierConstants.TopologyModule.FIELD_CONFIG, parser.parse(gson.toJson(topologyModule.getConfig())).getAsJsonObject());
+    	try {
```
Comment (amatosc) [inline sinfonier-web/app/models/topology/json/serializers/TopologySerializer.java]: Please, formatted correctly

**Comment by demonodojo** [comment]
Gson uses the name of the field to serialize it in json. We use xtype un
our topologies, so renaming it will do it fine.

**Comment by amatosc** [comment]
it's ok but you can also use @SerializedName from Gson.

File: sinfonier-web/app/models/module/ModuleVersion.java (hunk 1 above)
Comment (demonodojo) [reply sinfonier-web/app/models/module/ModuleVersion.java:278]: ok

File: sinfonier-web/app/models/responses/Codes.java (hunk 2 above)
Comment (demonodojo) [reply sinfonier-web/app/models/responses/Codes.java]: ok

File: sinfonier-web/app/models/topology/json/serializers/TopologySerializer.java (hunk 4 above)
Comment (demonodojo) [reply sinfonier-web/app/models/topology/json/serializers/TopologySerializer.java]: ok

File: sinfonier-web/app/models/topology/Wire.java (hunk 3 above)
Comment (demonodojo) [reply sinfonier-web/app/models/topology/Wire.java:11]: Json serializer uses the property name to generate keys. This key must be lowercase.

**Review by amatosc — CHANGES_REQUESTED** [review]
Please check the test some of them are failed 

~ BasicTest...                                   PASSED     1s
~ tests/BaseTest...                              PASSED     0s
~ tests/models/module/ModuleTest...              FAILED  !  0s
~ tests/models/module/ModuleVersionTest...       PASSED     0s
~ tests/models/storm/ParamsValidatorTests...     PASSED     0s
~ tests/models/topology/TopologyTests...         PASSED     0s
~ ApplicationTest...                             PASSED     2s
~ tests/BaseTestFunctional...                    PASSED     0s
~ Application...                                 PASSED     5s

For checking execute play autotest

**Comment by demonodojo** [comment]
A mi me aparecen los tests funcionando:

~ BasicTest...                                   PASSED     1s
~ tests/BaseTest...                              PASSED     0s
~ tests/models/module/ModuleTest...              PASSED     0s
~ tests/models/module/ModuleVersionTest...       PASSED     0s
~ tests/models/storm/ParamsValidatorTests...     PASSED     0s
~ tests/models/topology/TopologyTests...         PASSED     0s
~ ApplicationTest...                             PASSED     1s
~ tests/BaseTestFunctional...                    PASSED     0s
~ Application...                                 PASSED     4s

**Review by fandrod — APPROVED** [review]

File: sinfonier-web/app/models/topology/json/serializers/TopologySerializer.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+			jsonModule.add(models.SinfonierConstants.TopologyModule.FIELD_VALUES,
+					parser.parse(gson.toJson(topologyModule.getValues())).getAsJsonObject());
+			jsonModule.add(models.SinfonierConstants.TopologyModule.FIELD_CONFIG,
+					parser.parse(gson.toJson(topologyModule.getConfig())).getAsJsonObject());
+			try {
+				Module module = Module.findById(topologyModule.getModuleId());
+				ModuleVersion moduleVersion = ModuleVersion.findById(topologyModule.getModuleVersionId());
+				jsonModule.add(FIELD_MODULE, moduleVersion.toJson(module));
+			} catch (SinfonierException e) {
+				throw new JsonException(e.getMessage(), e.getCause());
```
Comment (amatosc) [inline sinfonier-web/app/models/topology/json/serializers/TopologySerializer.java:67]: We always raise SinfonierExceptions. I will retrieve my  request change. Maybe this not important but it is the style used. @ivan11paths choice.

**Review by amatosc — APPROVED** [review]

**Review by ivan11paths — APPROVED** [review]

File: sinfonier-web/app/models/topology/json/serializers/TopologySerializer.java (hunk 5 above)
Comment (demonodojo) [reply sinfonier-web/app/models/topology/json/serializers/TopologySerializer.java:67]: We are overriding a method that must implement an interface an cannot throw a SinfonierException

**Pull request merged**